File: src/Algorithms/Genetic Algorithm/replacement.js
function generationalReplacement(oldPopulation, offspring) {

    return offspring;
}

function elitismReplacement(
    oldPopulation,
    offspring,
    fitnessFunc,
    elitismCount = 1
) {

    const combinedPopulation = [...oldPopulation, ...offspring];
    combinedPopulation.sort((a, b) => fitnessFunc(b) - fitnessFunc(a));
    return combinedPopulation.slice(0, oldPopulation.length);
}

function steadyStateReplacement(oldPopulation, offspring, replacementCount) {

    const newPopulation = [...oldPopulation];
    for (let i = 0; i < replacementCount; i++) {
        if (i < offspring.length) {
            newPopulation[i] = offspring[i];
        }
    }
    return newPopulation;
} 

function tournamentReplacement(
    oldPopulation,
    offspring,
    fitnessFunc,
    tournamentSize = 3
) {

    const combinedPopulation = [...oldPopulation, ...offspring];
    const newPopulation = [];
    for (let i = 0; i < oldPopulation.length; i++) {
        const tournament = [];
        for (let j = 0; j < tournamentSize; j++) {
            const randomIndex = Math.floor(
                Math.random() * combinedPopulation.length
            );
            tournament.push(combinedPopulation[randomIndex]);
        }
        const winner = tournament.reduce(
            (best, individual) =>
                fitnessFunc(individual) > fitnessFunc(best) ? individual : best,
            tournament[0]
        );
        newPopulation.push(winner);
    }
    return newPopulation;
}

function fitnessBasedReplacement(oldPopulation, offspring, fitnessFunc) {

    const combinedPopulation = [...oldPopulation, ...offspring];
    combinedPopulation.sort((a, b) => fitnessFunc(b) - fitnessFunc(a));
    return combinedPopulation.slice(0, oldPopulation.length);
}

export function replacement(
    oldPopulation,
    offspring,
    fitnessFunc,
    method = "Generational",
    options = {}
) {
    switch (method) {
        case "Generational Replacement":
            return generationalReplacement(oldPopulation, offspring);
        case "Elitism":
            return elitismReplacement(
                oldPopulation,
                offspring,
                fitnessFunc,
                options.elitismCount
            );
        case "Steady State Replacement":
            return steadyStateReplacement(
                oldPopulation,
                offspring,
                options.replacementCount
            );
        case "Tournament Replacement":
            return tournamentReplacement(
                oldPopulation,
                offspring,
                fitnessFunc,
                options.tournamentSize
            );
        case "Fitness Based Replacement":
            return fitnessBasedReplacement(
                oldPopulation,
                offspring,
                fitnessFunc
            );
        default:
            throw new Error("Invalid replacement method");
    }
}

File: replacement.py
import random
from typing import List, Callable

# ========================
# 1. Generational Replacement
# ========================

def generational_replacement(old_population: List, offspring: List) -> List:
    """
    Replaces the entire old population with the offspring.
    """
    return offspring

# ========================
# 2. Elitism
# ========================

def elitism_replacement(old_population: List, offspring: List, fitness_func: Callable, elitism_count: int = 1) -> List:
    """
    Replaces the old population with the offspring but keeps the best `elitism_count` individuals from the old population.
    """
    combined_population = old_population + offspring
    combined_population.sort(key=fitness_func, reverse=True)
    return combined_population[:len(old_population)]

# ========================
# 3. Steady-State Replacement
# ========================

def steady_state_replacement(old_population: List, offspring: List, replacement_count: int) -> List:
    """
    Replaces only `replacement_count` individuals in the old population with the offspring.
    """
    new_population = old_population.copy()
    for i in range(replacement_count):
        if i < len(offspring):  # Ensure we don't exceed the offspring list
            new_population[i] = offspring[i]
    return new_population

# ========================
# 4. Tournament Replacement
# ========================

def tournament_replacement(old_population: List, offspring: List, fitness_func: Callable, tournament_size: int = 3) -> List:
    """
    Uses tournament selection to decide which individuals (from old population and offspring) survive to the next generation.
    """
    combined_population = old_population + offspring
    new_population = []
    for _ in range(len(old_population)):
        tournament = random.sample(combined_population, tournament_size)
        winner = max(tournament, key=fitness_func)
        new_population.append(winner)
    return new_population

# ========================
# 5. Fitness-Based Replacement
# ========================

def fitness_based_replacement(old_population: List, offspring: List, fitness_func: Callable) -> List:
    """
    Replaces the worst individuals in the old population with the offspring.
    """
    combined_population = old_population + offspring
    combined_population.sort(key=fitness_func, reverse=True)
    return combined_population[:len(old_population)]

# ========================
# 6. Replacement Wrapper
# ========================

def replace(
    old_population: List,
    offspring: List,
    method: str,
    fitness_func: Callable = None,
    **kwargs
) -> List:
    """
    Wrapper function to call the appropriate replacement method.
    """
    if method == "generational":
        return generational_replacement(old_population, offspring)
    elif method == "elitism":
        return elitism_replacement(old_population, offspring, fitness_func, kwargs.get("elitism_count", 1))
    elif method == "steady_state":
        return steady_state_replacement(old_population, offspring, kwargs.get("replacement_count", 1))
    elif method == "tournament":
        return tournament_replacement(old_population, offspring, fitness_func, kwargs.get("tournament_size", 3))
    elif method == "fitness_based":
        return fitness_based_replacement(old_population, offspring, fitness_func)
    else:
        raise ValueError(f"Unknown replacement method: {method}")